llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -1987,11 +1979,47 @@ MemoryDepChecker::getDependenceDistanceStrideAndSize(
     }
   }
 
-  // Need accesses with constant strides and the same direction. We don't want
-  // to vectorize "A[B[i]] += ..." and similar code or pointer arithmetic that
-  // could wrap in the address space.
-  if (!StrideAPtr || !StrideBPtr || (StrideAPtr > 0 && StrideBPtr < 0) ||
-      (StrideAPtr < 0 && StrideBPtr > 0)) {
+  // Need accesses with constant strides and the same direction for further
+  // dependence analysis. We don't want to vectorize "A[B[i]] += ..." and
+  // similar code or pointer arithmetic that could wrap in the address space.
+  //
+  // If Src or Sink are non-wrapping AddRecs, StrideAPtr and StrideBPtr contain
+  // a SCEV representing the stride of the SCEV. It may not be a known constant
+  // value though.
+
+  // If either Src or Sink are not strided (i.e. not a non-wrapping AddRec), we
+  // cannot analyze the dependence further.
+  if (!StrideAPtr || !StrideBPtr) {
+    bool SrcInvariant = SE.isLoopInvariant(Src, InnermostLoop);
+    bool SinkInvariant = SE.isLoopInvariant(Sink, InnermostLoop);
+    // If either Src or Sink are not loop invariant and not strided, the
+    // expression in the current iteration may overlap with any earlier or later
+    // iteration. This is not safe and we also cannot generate runtime checks to
+    // ensure safety. This includes expressions where an index is loaded in each
+    // iteration or wrapping AddRecs.
+    if ((!SrcInvariant && !StrideAPtr) || (!SinkInvariant && !StrideBPtr))
+      return MemoryDepChecker::Dependence::IndirectUnsafe;
```

Comment:
> Agreed, naming could be improved separately

Also, `IndirectUnsafe` may have been intended to only apply for the `A[B[i]]` case (access ptr depends on index loaded from memory whose address itself is not invariant). https://github.com/llvm/llvm-project/issues/87189 may have demonstrated that trying to identify the cases that don't work is bound to fail (in this case forgetting other SCEVUnknown can be source of unpredictable access as well, for instance the return value of a function call), but instead one should identify the cases that are known to be supported (in this case: constant strides). That means `IndirectUnsafe` should be renamed as well, since it doesn't just apply to `A[B[i]]` anymore.

You probably intend to extend `getPtrStride` to return a SCEV so one can support non-constant strides, but that are invariant to the loop. It might be useful to pass APtr/BPtr as a SCEV only (`Src`/`Sink`) consistently as well instead calling `PSE.getSCEV` on-demand.

Reply:
Will look into renaming separately, thanks!

> You probably intend to extend getPtrStride to return a SCEV so one can support non-constant strides, but that are invariant to the loop. It might be useful to pass APtr/BPtr as a SCEV only (Src/Sink) consistently as well instead calling PSE.getSCEV on-demand.

Yep, will also work on that as follow-up, thanks!